builder/main.py

```diff
@@ -109,6 +115,103 @@ def generate_uf2(target, source, env):
+        "M": 1024*1024,
+        "MB": 1024*1024,
+        "K": 1024,
+        "KB": 1024,
+        "B": 1,
+        "": 1 # giving no conversion factor is factor 1.
+    }
+    # match <floating pointer number><conversion factor>.
+    extract_regex = r'^((?:[0-9]*[.])?[0-9]+)([mkbMKB]*)$'
+    res = re.findall(extract_regex, expression)
+    # unparsable expression? Warning.
+    if len(res) == 0:
+        sys.stderr.write(
+            "Error: Could not parse filesystem size expression '%s'."
+            " Will treat as size = 0.\n" % str(expression))
+        return 0
+    # access first result
+    number, factor = res[0]
+    number = float(number)
+    number *= conversion_factors[factor.upper()]
+    return int(number)
+
+def fetch_fs_size(env):
+    # follow generation formulas from makeboards.py for Earle Philhower core
+    # given the total flash size, a user can specify
+    # the amount for the filesystem (0MB, 2MB, 4MB, 8MB, 16MB)
+    # via board_build.filesystem_size,
+    # and we will calculate the flash size and eeprom size from that.
+    flash_size = board.get("upload.maximum_size")
+    filesystem_size = board.get("build.filesystem_size", "0MB")
+    filesystem_size_int = convert_size_expression_to_int(filesystem_size)
+    # last 4K are allocated for EEPROM emulation in flash.
+    # see https://github.com/earlephilhower/arduino-pico/blob/3414b73172d307e9dc901f7fee83b41112f73457/libraries/EEPROM/EEPROM.cpp#L43-L46
+    eeprom_size = 4096
+
+    maximum_sketch_size = flash_size - eeprom_size - filesystem_size_int
+
+    print("Flash size: %.2fMB" % (flash_size / 1024.0 / 1024.0))
+    print("Sketch size: %.2fMB" % (maximum_sketch_size / 1024.0 / 1024.0))
+    print("Filesystem size: %.2fMB" % (filesystem_size_int / 1024.0 / 1024.0))
+
+    eeprom_start = 0x10000000 + flash_size - eeprom_size
+    fs_start = 0x10000000 + flash_size - eeprom_size - filesystem_size_int
+    fs_end = 0x10000000 + flash_size - eeprom_size
+
+    if maximum_sketch_size <= 0:
+        sys.stderr.write(
+            "Error: Filesystem too large for given flash. "
+            "Can at max be flash size - 4096 bytes. "
+            "Available sketch size with current "
+            "config would be %d bytes.\n" % maximum_sketch_size)
+        sys.stderr.flush()
+        env.Exit(1)
+
+    env["PICO_FLASH_LENGTH"] = maximum_sketch_size
+    env["PICO_EEPROM_START"] = eeprom_start
+    env["FS_START"] = fs_start
+    env["FS_END"] = fs_end
+    # LittleFS configuration parameters taken from
+    # https://github.com/earlephilhower/arduino-pico-littlefs-plugin/blob/master/src/PicoLittleFS.java
+    env["FS_PAGE"] = 256
+    env["FS_BLOCK"] = 4096
+
+    print("Maximium Sketch size: %d "
+        "EEPROM start: %s Filesystem start: %s "
+        "Filesystem end: %s" % 
+        (maximum_sketch_size, hex(eeprom_start), hex(fs_start), hex(fs_end)))
+
+
+def __fetch_fs_size(target, source, env):
+    fetch_fs_size(env)
+    return (target, source)
```

Comment:
Just a tiny improvement, could you please move your helper functions `fetch_fs_size`,  `convert_size_expression_to_int`, etc. to the top of the file and then add the `__fetch_fs_size` to the env in the `Replace` section above? You can take a look how it's structured in the espressif32 repo https://github.com/platformio/platform-espressif32/blob/develop/builder/main.py#L173